src/dex/dex.utils.cpp

```diff
@@ -0,0 +1,85 @@
+/******************************************************************************
+ * Copyright © 2013-2019 The Komodo Platform Developers.                      *
+ *                                                                            *
+ * See the AUTHORS, DEVELOPER-AGREEMENT and LICENSE files at                  *
+ * the top-level directory of this distribution for the individual copyright  *
+ * holder information and the developer policies on copyright and licensing.  *
+ *                                                                            *
+ * Unless otherwise agreed in a custom licensing agreement, no part of the    *
+ * Komodo Platform software, including this file may be copied, modified,     *
+ * propagated or distributed except according to the terms contained in the   *
+ * LICENSE file                                                               *
+ *                                                                            *
+ * Removal or modification of this copyright notice is prohibited.            *
+ *                                                                            *
+ ******************************************************************************/
+
+#include <vector>
+#include <unordered_set>
+#include <unordered_map>
+#include <string>
+
+#include <absl/numeric/int128.h>
+#include <orders/orders.hpp>
+#include <utils/mmbot_strong_types.hpp>
+#include <utils/antara.utils.hpp>
+#include <mm2/mm2.client.hpp>
+
+#include "dex.utils.hpp"
+
+namespace antara::mmbot
+{
+    mm2::buy_request to_buy (orders::order_level ol, antara::pair pair)
+    {
+        const auto& cfg = get_mmbot_config();
+
+        auto base = pair.base;
+        auto quote = pair.quote;
+        auto price = get_price_as_string_decimal(cfg, base.symbol, base.symbol, ol.price);
+        auto quantity = std::to_string(ol.quantity.value());
+
+        return mm2::buy_request{base, quote, price, quantity};
+    }
+
+    mm2::sell_request to_sell (orders::order_level ol, antara::pair pair)
+    {
+        const auto& cfg = get_mmbot_config();
+
+        auto base = pair.base;
+        auto quote = pair.quote;
+        auto price = get_price_as_string_decimal(cfg, base.symbol, base.symbol, ol.price);
+        auto quantity = std::to_string(ol.quantity.value());
+
+        return mm2::sell_request{base, quote, price, quantity};
+    }
```

Comment:
I'm thinking about a nice way to abstract this code, keep it DRY. If I'm right, we can't cast a struct, so that rules out having a function that returns a `trade_request` that gets cast.

@Milerius, do you have any ideas?

Reply:
you can simply use template ?

```cpp
template <typename TRequest>
TRequest foo (orders::order_level ol, antara::pair pair)
    {
        const auto& cfg = get_mmbot_config();

        auto base = pair.base;
        auto quote = pair.quote;
        auto price = get_price_as_string_decimal(cfg, base.symbol, base.symbol, ol.price);
        auto quantity = std::to_string(ol.quantity.value());

        return TRequest{base, quote, price, quantity};
    }

foo<order::buy>(...);
foo<order::sell>();
```

Reply:
(i'm still thinking that we can refactor that, let me know if you want me to do it my friend !)